pkg/etcd/client.go

```diff
@@ -313,3 +322,229 @@ func isRetryableError(rpcName string) retry.IsRetryable {
 		return true
 	}
 }
+
+// The following code is mainly copied from:
+// https://github.com/tikv/pd/blob/master/pkg/utils/etcdutil/etcdutil.go
+const (
+	// defaultEtcdClientTimeout is the default timeout for etcd client.
+	defaultEtcdClientTimeout = 5 * time.Second
+	// defaultDialKeepAliveTime is the time after which client pings the server to see if transport is alive.
+	defaultDialKeepAliveTime = 10 * time.Second
+	// defaultDialKeepAliveTimeout is the time that the client waits for a response for the
+	// keep-alive probe. If the response is not received in this time, the connection is closed.
+	defaultDialKeepAliveTimeout = 3 * time.Second
+	// etcdServerOfflineTimeout is the timeout for an unhealthy etcd endpoint to be offline from healthy checker.
+	etcdServerOfflineTimeout = 30 * time.Minute
+	// etcdServerDisconnectedTimeout is the timeout for an unhealthy etcd endpoint to be disconnected from healthy checker.
+	etcdServerDisconnectedTimeout = 1 * time.Minute
+	// healthyPath is the path to check etcd health.
+	healthyPath = "health"
+)
+
+func newClient(tlsConfig *tls.Config, endpoints ...string) (*clientv3.Client, error) {
+	if len(endpoints) == 0 {
+		return nil, errors.New("empty endpoints")
+	}
+	logConfig := logutil.DefaultZapLoggerConfig
+	logConfig.Level = zap.NewAtomicLevelAt(zapcore.ErrorLevel)
+
+	lgc := zap.NewProductionConfig()
+	lgc.Encoding = log.ZapEncodingName
+	client, err := clientv3.New(clientv3.Config{
+		Endpoints:            endpoints,
+		TLS:                  tlsConfig,
+		LogConfig:            &logConfig,
+		DialTimeout:          defaultEtcdClientTimeout,
+		DialKeepAliveTime:    defaultDialKeepAliveTime,
+		DialKeepAliveTimeout: defaultDialKeepAliveTimeout,
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return client, err
+}
+
+// CreateRawEtcdClient creates etcd v3 client with detecting endpoints.
+// It will check the health of endpoints periodically, and update endpoints if needed.
+func CreateRawEtcdClient(tlsConfig *tls.Config, endpoints ...string) (*clientv3.Client, error) {
+	client, err := newClient(tlsConfig, endpoints...)
+	if err != nil {
+		return nil, err
+	}
+
+	tickerInterval := defaultDialKeepAliveTime
+
+	checker := &healthyChecker{
+		tlsConfig: tlsConfig,
+	}
+	eps := syncUrls(client)
+	checker.update(eps)
+
+	// Create a goroutine to check the health of etcd endpoints periodically.
+	go func(client *clientv3.Client) {
```

Comment:
this spawns 2 goroutines which are not managed by any synchronization mechanism, such as the `WaitGroup`

Reply:
These two goroutines have the same lifecycle as the CDC server process, so there is no need to add any additional controls for them.